### PR TITLE
added homo fm

### LINKED ISSUE #961: support HomoFM Algorithm
In FATE version >= 1.3, Homo FM will be support.

### PR DESCRIPTION
**Pull request by versusniubility**
Fixes  ISSUE #xxx

Changes:

1. added HOMO FM algorithm

2. added decay checking

3. fix bugs

Closes #961

### CODE REVIEW

File: federatedml/transfer_variable/definition/transfer_conf.json [hunk 1]
```diff
@@ -1188,108 +1327,5 @@
     "guest"
    ]
   }
- },
- "HeteroGMFTransferVariable": {
```
Comment (mgqa34) [inline federatedml/transfer_variable/definition/transfer_conf.json]: This line delete the transfer variable definition of GMF，is this ok？

File: federatedrec/factorization_machine/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    a)	“sgd”: gradient descent with arbitrary batch size
+    b) “rmsprop”: RMSProp
+    c) “adam”: Adam
+    d) “adagrad”: AdaGrad
+    e) “nesterov_momentum_sgd”: Nesterov Momentum
+4. Three converge criteria:
+    a) "diff": Use difference of loss between two iterations, not available for multi-host training
+    b) "abs": Use the absolute value of loss
+    c) "weight_diff": Use difference of model weights
+5. Support multi-host modeling task. For details on how to configure for multi-host modeling task, please refer to this [guide](../../../doc/dsl_conf_setting_guide.md)
```
Comment (mgqa34) [inline federatedrec/factorization_machine/README.md]: This section should be homo-fm? In previous pull request, HeteroFM does not support multi-host

File: federatedrec/factorization_machine/homo_factorization_machine/homo_fm_base.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    def _init_model(self, params):
+        super(HomoFMBase, self)._init_model(params)
+        self.re_encrypt_batches = params.re_encrypt_batches
+
+        # if params.encrypt_param.method == consts.PAILLIER:
+        #     self.cipher_operator = PaillierEncrypt()
+        # else:
+        #     self.cipher_operator = FakeEncrypt()
+
+        self.cipher_operator = FakeEncrypt()
```
Comment (mgqa34) [inline federatedrec/factorization_machine/homo_factorization_machine/homo_fm_base.py]: The code above may be recovery for using paillier.

File: federatedml/transfer_variable/definition/transfer_conf.json [hunk 4]
```diff
@@ -875,6 +875,145 @@
...
+   ]
+  },
+  "aggregated_model": {
+   "src": "arbiter",
+   "dst": [
+    "guest",
+    "host"
+   ]
+  },
+  "to_encrypt_model": {
```
Comment (mgqa34) [inline federatedml/transfer_variable/definition/transfer_conf.json]: unused transfer variable should delete, like "pailler_pubkey"\"to_encrypt_model"\"re_encrypted_model"

File: federatedrec/factorization_machine/homo_factorization_machine/homo_fm_arbiter.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+            validation_strategy.validate(self, self.n_iter_)
+            self.n_iter_ += 1
+
+        LOGGER.info("Finish Training task, total iters: {}".format(self.n_iter_))
+
+    def predict(self, data_instantces=None):
+        LOGGER.info(f'Start predict task')
+
+    def run(self, component_parameters=None, args=None):
```
Comment (tanmc123) [inline federatedrec/factorization_machine/homo_factorization_machine/homo_fm_arbiter.py]: The run function is no need to be rewrite in arbiter now. Suggest to remove and test it again.

File: federatedrec/factorization_machine/homo_factorization_machine/homo_fm_guest.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        mini_batch_obj = MiniBatch(data_inst=data_instances, batch_size=self.batch_size)
+        model_weights = self.model_weights
+
+        degree = 0
+        while self.n_iter_ < self.max_iter:
+            LOGGER.info("iter:{}".format(self.n_iter_))
+            batch_data_generator = mini_batch_obj.mini_batch_data_generator()
+
+            self.optimizer.set_iters(self.n_iter_)
+            if self.n_iter_ > 0 and self.n_iter_ % self.aggregate_iters == 0:
```
Comment (tanmc123) [inline federatedrec/factorization_machine/homo_factorization_machine/homo_fm_guest.py]: This one make an aggregation if True. However, after updated last iteration, the updated models are not aggregated. Suggest to add a final aggregation in the last iteration.

File: federatedrec/factorization_machine/homo_factorization_machine/homo_fm_host.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+        mini_batch_obj = MiniBatch(data_inst=data_instances, batch_size=self.batch_size)
+
+        model_weights = self.model_weights
+        degree = 0
+        while self.n_iter_ < self.max_iter:
+            LOGGER.info("iter:{}".format(self.n_iter_))
+            batch_data_generator = mini_batch_obj.mini_batch_data_generator()
+
+            if self.n_iter_ > 0 and self.n_iter_ % self.aggregate_iters == 0:
```
Comment (tanmc123) [inline federatedrec/factorization_machine/homo_factorization_machine/homo_fm_host.py]: Same problem with the same part in guest

File: federatedrec/factorization_machine/homo_factorization_machine/homo_fm_arbiter.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        max_iter = self.max_iter
+        validation_strategy = self.init_validation_strategy()
+
+        model_shape = -1
+        embed_size = self.init_param_obj.embed_size
+
+        while self.n_iter_ < max_iter:
+            suffix = (self.n_iter_,)
+
+            if self.n_iter_ > 0 and self.n_iter_ % self.aggregate_iters == 0:
```
Comment (tanmc123) [inline federatedrec/factorization_machine/homo_factorization_machine/homo_fm_arbiter.py]: Same problem with the same part in guest

File: federatedrec/optim/gradient/homo_fm_gradient.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+
+class FactorizationGradient(object):
+
+    @staticmethod
+    def compute_loss(values, w, embed, intercept):
+        LOGGER.info("compute loss")
+        X, Y = load_data(values)
+        fm = fm_func(X, embed)
+        tot_loss = np.log(1 + np.exp(np.multiply(-Y.transpose(), X.dot(w) + fm + intercept))).sum()
```
Comment (tanmc123) [inline federatedrec/optim/gradient/homo_fm_gradient.py]: This formular can be simplified as log(1 + exp(x)), where x is something like -ywx + fm + intercept.
However, when x is relatively large like larger than 710, exp(x) will get inf which cause tot_loss inf. It is better to use log(1+exp(x)) = x if x is large enough to avoid this. 

Fate has provided a function in optim/activation called log_logistic. log(1+exp(x)) can safely be calculated as -log_logistic(-x) to solve this problem.

File: federatedrec/param/factorization_machine_param.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+            self.penalty = self.penalty.upper()
+            if self.penalty not in [consts.L1_PENALTY, consts.L2_PENALTY, 'NONE']:
+                raise ValueError(
+                    "factorization_param's penalty not supported, penalty should be 'L1', 'L2' or 'none'")
+
+        if not isinstance(self.tol, (int, float)):
+            raise ValueError(
+                "factorization_param's tol {} not supported, should be float type".format(self.tol))
+
+        if not isinstance(self.clip_gradient, (int, float, None)):
```
Comment (tanmc123) [inline federatedrec/param/factorization_machine_param.py]: If self.clip_gradient is None, this line will report an error.

File: federatedrec/param/factorization_machine_param.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+                raise ValueError(
+                    "factorization_param's early_stop not supported, converge_func should be"
+                    " 'diff' or 'abs'")
+
+        self.encrypt_param.check()
+        if self.encrypt_param.method not in [consts.PAILLIER, None]:
+            raise ValueError(
+                "factorization_param's encrypted method support 'Paillier' or None only")
+
+        if type(self.decay).__name__ not in ["int", 'float']:
```
Comment (tanmc123) [inline federatedrec/param/factorization_machine_param.py]: It is better to add a judgement of whether decay is smaller than 0 or not since gradient of fm is easy to explode with the growth of learning_rate

File: federatedrec/transfer_variable/definition/homo_fm.json [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+      ]
+    },
+    "dh_ciphertext_bc": {
+      "src": "arbiter",
+      "dst": [
+        "guest",
+        "host"
+      ]
+    },
+    "paillier_pubkey": {
```
Comment (tanmc123) [inline federatedrec/transfer_variable/definition/homo_fm.json]: If Paillier encryption is no needed, we suggest to remove transfer_variable relative with it to avoid potential misunderstanding.

File: federatedml/transfer_variable/definition/transfer_conf.json (hunk 1 above)
Comment (versusniubility) [reply federatedml/transfer_variable/definition/transfer_conf.json]: the transfer_variable folder is under federatedrec, so this part can ignor
